matchmake-extension/auto_matchmake_postpone.go

```diff
@@ -0,0 +1,102 @@
+package matchmake_extension
+
+import (
+	"math"
+
+	"github.com/PretendoNetwork/mario-kart-7-secure/globals"
+	nex "github.com/PretendoNetwork/nex-go"
+	match_making "github.com/PretendoNetwork/nex-protocols-go/match-making"
+	matchmake_extension "github.com/PretendoNetwork/nex-protocols-go/matchmake-extension"
+	"github.com/PretendoNetwork/nex-protocols-go/notifications"
+	common_globals "github.com/PretendoNetwork/nex-protocols-common-go/globals"
+	"encoding/json"
+)
+
+func AutoMatchmake_Postpone(err error, client *nex.Client, callID uint32, matchmakeSession *match_making.MatchmakeSession, message string) {
+	missingHandler := false
+	if commonMatchmakeExtensionProtocol.CleanupSearchMatchmakeSessionHandler == nil {
+		logger.Warning("MatchmakeExtension::AutoMatchmake_Postpone missing CleanupSearchMatchmakeSessionHandler!")
+		missingHandler = true
+	}
+	if missingHandler {
+		return
+	}
+	//This is the best way I found to copy the full object. And I still hate it.
+	tmp, _ := json.Marshal(matchmakeSession)
+	matchmakeSessionCopy := match_making.NewMatchmakeSession()
+	json.Unmarshal(tmp, &matchmakeSessionCopy)
+	searchMatchmakeSession := commonMatchmakeExtensionProtocol.CleanupSearchMatchmakeSessionHandler(*matchmakeSessionCopy)
+	sessionIndex := FindSearchMatchmakeSession(searchMatchmakeSession)
+	if sessionIndex == math.MaxUint32 {
+		session := common_globals.CommonMatchmakeSession{
+			SearchMatchmakeSession: searchMatchmakeSession,
+			GameMatchmakeSession:   *matchmakeSession,
+		}
+		sessionIndex = len(common_globals.Sessions)
+		common_globals.Sessions = append(common_globals.Sessions, session)
+		common_globals.Sessions[sessionIndex].GameMatchmakeSession.Gathering.ID = uint32(sessionIndex)
+		common_globals.Sessions[sessionIndex].GameMatchmakeSession.Gathering.OwnerPID = client.PID()
+		common_globals.Sessions[sessionIndex].GameMatchmakeSession.Gathering.HostPID = client.PID()
+	}
+
+	common_globals.Sessions[sessionIndex].PlayersByConnectionId = append(common_globals.Sessions[sessionIndex].PlayersByConnectionId, client.ConnectionID())
+
+	rmcResponseStream := nex.NewStreamOut(globals.NEXServer)
+	rmcResponseStream.WriteString("MatchmakeSession")
+	lengthStream := nex.NewStreamOut(globals.NEXServer)
+	lengthStream.WriteStructure(&common_globals.Sessions[sessionIndex].GameMatchmakeSession)
+	matchmakeSessionLength := uint32(len(lengthStream.Bytes()))
+	rmcResponseStream.WriteUInt32LE(matchmakeSessionLength + 4)
+	rmcResponseStream.WriteUInt32LE(matchmakeSessionLength)
+	rmcResponseStream.WriteStructure(&common_globals.Sessions[sessionIndex].GameMatchmakeSession)
```

Comment:
What is happening here? This looks like you're manually writing the data for a DataHolder type? If so, `StreamOut` has a `WriteDataHolder` method made for this very task

Also when looking at Kinnay's wiki the DataHolder type is `Gathering` but you are writing a `MatchmakeSession`?